src/inmanta/util/collections.py

```diff
@@ -0,0 +1,167 @@
+"""
+    Copyright 2024 Inmanta
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+    Contact: code@inmanta.com
+"""
+
+from typing import Generic, Optional, Self, TypeVar
+from collections.abc import Hashable, Iterator, MutableMapping, Set
+
+
+P = TypeVar("P", bound=Hashable)
+S = TypeVar("S", bound=Hashable)
+# type vars not bound to the class
+K = TypeVar("K", bound=Hashable)
+V = TypeVar("V", bound=Hashable)
+
+
+# TODO: review
+# TODO: unit tests
+
+
+# TODO: better name?
+class BidirectionalManyToManyMapping(MutableMapping[P, Set[S]], Generic[P, S]):
```

Comment:
Cool! 

(If we are very looking for a candidate to turn into C or or Rust, might be this one)